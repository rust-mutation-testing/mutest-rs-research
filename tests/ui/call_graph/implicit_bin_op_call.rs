//@ print-call-graph
//@ print-targets
//@ stdout
//@ stderr: empty

#![feature(decl_macro)]

macro test($test:ident, $ty:ident, $op_trait:path|$op_assign_trait:path, $op_fn:ident|$op_assign_fn:ident, $op_impl_fn:ident|$op_assign_impl_fn:ident, $op:tt|$op_assign:tt) {
    struct $ty;

    fn $op_impl_fn() {}
    impl $op_trait for $ty {
        type Output = Self;

        fn $op_fn(self, _rhs: Self) -> Self::Output {
            $op_impl_fn();
            Self
        }
    }

    fn $op_assign_impl_fn() {}
    impl $op_assign_trait for $ty {
        fn $op_assign_fn(&mut self, _rhs: Self) {
            $op_assign_impl_fn();
            *self = Self;
        }
    }

    #[test]
    fn $test() {
        let mut v = $ty $op $ty;
        v $op_assign $ty;
    }
}

test!(test_implicit_add_call, ImplsAdd, ::std::ops::Add|::std::ops::AddAssign, add|add_assign, add_impl|add_assign_impl, + | +=);
test!(test_implicit_sub_call, ImplsSub, ::std::ops::Sub|::std::ops::SubAssign, sub|sub_assign, sub_impl|sub_assign_impl, - | -=);
test!(test_implicit_mul_call, ImplsMul, ::std::ops::Mul|::std::ops::MulAssign, mul|mul_assign, mul_impl|mul_assign_impl, * | *=);
test!(test_implicit_div_call, ImplsDiv, ::std::ops::Div|::std::ops::DivAssign, div|div_assign, div_impl|div_assign_impl, / | /=);
test!(test_implicit_rem_call, ImplsRem, ::std::ops::Rem|::std::ops::RemAssign, rem|rem_assign, rem_impl|rem_assign_impl, % | %=);
test!(test_implicit_bit_or_call, ImplsBitOr, ::std::ops::BitOr|::std::ops::BitOrAssign, bitor|bitor_assign, bitor_impl|bitor_assign_impl, | | |=);
test!(test_implicit_bit_xor_call, ImplsBitXor, ::std::ops::BitXor|::std::ops::BitXorAssign, bitxor|bitxor_assign, bitxor_impl|bitxor_assign_impl, ^ | ^=);
test!(test_implicit_bit_and_call, ImplsBitAnd, ::std::ops::BitAnd|::std::ops::BitAndAssign, bitand|bitand_assign, bitand_impl|bitand_assign_impl, & | &=);
test!(test_implicit_shl_call, ImplsShl, ::std::ops::Shl|::std::ops::ShlAssign, shl|shl_assign, shl_impl|shl_assign_impl, << | <<=);
test!(test_implicit_shr_call, ImplsShr, ::std::ops::Shr|::std::ops::ShrAssign, shr|shr_assign, shr_impl|shr_assign_impl, >> | >>=);
