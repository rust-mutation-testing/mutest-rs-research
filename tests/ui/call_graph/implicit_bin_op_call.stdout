
@@@ call graph @@@

entry points:

test test_implicit_add_call
  -> <ImplsAdd as std::ops::Add>::add at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#4)
  -> <ImplsAdd as std::ops::AddAssign>::add_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#4)
test test_implicit_bit_and_call
  -> <ImplsBitAnd as std::ops::BitAnd>::bitand at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#18)
  -> <ImplsBitAnd as std::ops::BitAndAssign>::bitand_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#18)
test test_implicit_bit_or_call
  -> <ImplsBitOr as std::ops::BitOr>::bitor at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#14)
  -> <ImplsBitOr as std::ops::BitOrAssign>::bitor_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#14)
test test_implicit_bit_xor_call
  -> <ImplsBitXor as std::ops::BitXor>::bitxor at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#16)
  -> <ImplsBitXor as std::ops::BitXorAssign>::bitxor_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#16)
test test_implicit_div_call
  -> <ImplsDiv as std::ops::Div>::div at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#10)
  -> <ImplsDiv as std::ops::DivAssign>::div_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#10)
test test_implicit_mul_call
  -> <ImplsMul as std::ops::Mul>::mul at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#8)
  -> <ImplsMul as std::ops::MulAssign>::mul_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#8)
test test_implicit_rem_call
  -> <ImplsRem as std::ops::Rem>::rem at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#12)
  -> <ImplsRem as std::ops::RemAssign>::rem_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#12)
test test_implicit_shl_call
  -> <ImplsShl as std::ops::Shl>::shl at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#20)
  -> <ImplsShl as std::ops::ShlAssign>::shl_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#20)
test test_implicit_shr_call
  -> <ImplsShr as std::ops::Shr>::shr at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#22)
  -> <ImplsShr as std::ops::ShrAssign>::shr_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#22)
test test_implicit_sub_call
  -> <ImplsSub as std::ops::Sub>::sub at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#6)
  -> <ImplsSub as std::ops::SubAssign>::sub_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#6)

nested calls at distance 1:

<ImplsAdd as std::ops::Add>::add at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#4)
  -> add_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#4)
<ImplsSub as std::ops::Sub>::sub at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#6)
  -> sub_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#6)
<ImplsMul as std::ops::Mul>::mul at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#8)
  -> mul_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#8)
<ImplsDiv as std::ops::Div>::div at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#10)
  -> div_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#10)
<ImplsRem as std::ops::Rem>::rem at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#12)
  -> rem_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#12)
<ImplsBitOr as std::ops::BitOr>::bitor at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#14)
  -> bitor_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#14)
<ImplsBitXor as std::ops::BitXor>::bitxor at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#16)
  -> bitxor_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#16)
<ImplsBitAnd as std::ops::BitAnd>::bitand at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#18)
  -> bitand_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#18)
<ImplsShl as std::ops::Shl>::shl at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#20)
  -> shl_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#20)
<ImplsShr as std::ops::Shr>::shr at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#22)
  -> shr_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#22)
<ImplsAdd as std::ops::AddAssign>::add_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#4)
  -> add_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#4)
<ImplsSub as std::ops::SubAssign>::sub_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#6)
  -> sub_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#6)
<ImplsMul as std::ops::MulAssign>::mul_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#8)
  -> mul_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#8)
<ImplsDiv as std::ops::DivAssign>::div_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#10)
  -> div_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#10)
<ImplsRem as std::ops::RemAssign>::rem_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#12)
  -> rem_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#12)
<ImplsBitOr as std::ops::BitOrAssign>::bitor_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#14)
  -> bitor_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#14)
<ImplsBitXor as std::ops::BitXorAssign>::bitxor_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#16)
  -> bitxor_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#16)
<ImplsBitAnd as std::ops::BitAndAssign>::bitand_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#18)
  -> bitand_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#18)
<ImplsShl as std::ops::ShlAssign>::shl_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#20)
  -> shl_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#20)
<ImplsShr as std::ops::ShrAssign>::shr_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#22)
  -> shr_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#22)

@@@ targets @@@

tests -(1)-> add_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#4)
  (1) test_implicit_add_call

tests -(1)-> sub_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#6)
  (1) test_implicit_sub_call

tests -(1)-> mul_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#8)
  (1) test_implicit_mul_call

tests -(1)-> div_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#10)
  (1) test_implicit_div_call

tests -(1)-> rem_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#12)
  (1) test_implicit_rem_call

tests -(1)-> bitor_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#14)
  (1) test_implicit_bit_or_call

tests -(1)-> bitxor_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#16)
  (1) test_implicit_bit_xor_call

tests -(1)-> bitand_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#18)
  (1) test_implicit_bit_and_call

tests -(1)-> shl_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#20)
  (1) test_implicit_shl_call

tests -(1)-> shr_impl at tests/ui/call_graph/implicit_bin_op_call.rs:11:5: 11:21 (#22)
  (1) test_implicit_shr_call

tests -(0)-> <ImplsAdd as std::ops::Add>::add at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#4)
  (0) test_implicit_add_call

tests -(0)-> <ImplsSub as std::ops::Sub>::sub at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#6)
  (0) test_implicit_sub_call

tests -(0)-> <ImplsMul as std::ops::Mul>::mul at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#8)
  (0) test_implicit_mul_call

tests -(0)-> <ImplsDiv as std::ops::Div>::div at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#10)
  (0) test_implicit_div_call

tests -(0)-> <ImplsRem as std::ops::Rem>::rem at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#12)
  (0) test_implicit_rem_call

tests -(0)-> <ImplsBitOr as std::ops::BitOr>::bitor at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#14)
  (0) test_implicit_bit_or_call

tests -(0)-> <ImplsBitXor as std::ops::BitXor>::bitxor at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#16)
  (0) test_implicit_bit_xor_call

tests -(0)-> <ImplsBitAnd as std::ops::BitAnd>::bitand at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#18)
  (0) test_implicit_bit_and_call

tests -(0)-> <ImplsShl as std::ops::Shl>::shl at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#20)
  (0) test_implicit_shl_call

tests -(0)-> <ImplsShr as std::ops::Shr>::shr at tests/ui/call_graph/implicit_bin_op_call.rs:15:9: 15:52 (#22)
  (0) test_implicit_shr_call

tests -(1)-> add_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#4)
  (1) test_implicit_add_call

tests -(1)-> sub_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#6)
  (1) test_implicit_sub_call

tests -(1)-> mul_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#8)
  (1) test_implicit_mul_call

tests -(1)-> div_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#10)
  (1) test_implicit_div_call

tests -(1)-> rem_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#12)
  (1) test_implicit_rem_call

tests -(1)-> bitor_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#14)
  (1) test_implicit_bit_or_call

tests -(1)-> bitxor_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#16)
  (1) test_implicit_bit_xor_call

tests -(1)-> bitand_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#18)
  (1) test_implicit_bit_and_call

tests -(1)-> shl_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#20)
  (1) test_implicit_shl_call

tests -(1)-> shr_assign_impl at tests/ui/call_graph/implicit_bin_op_call.rs:21:5: 21:28 (#22)
  (1) test_implicit_shr_call

tests -(0)-> <ImplsAdd as std::ops::AddAssign>::add_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#4)
  (0) test_implicit_add_call

tests -(0)-> <ImplsSub as std::ops::SubAssign>::sub_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#6)
  (0) test_implicit_sub_call

tests -(0)-> <ImplsMul as std::ops::MulAssign>::mul_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#8)
  (0) test_implicit_mul_call

tests -(0)-> <ImplsDiv as std::ops::DivAssign>::div_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#10)
  (0) test_implicit_div_call

tests -(0)-> <ImplsRem as std::ops::RemAssign>::rem_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#12)
  (0) test_implicit_rem_call

tests -(0)-> <ImplsBitOr as std::ops::BitOrAssign>::bitor_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#14)
  (0) test_implicit_bit_or_call

tests -(0)-> <ImplsBitXor as std::ops::BitXorAssign>::bitxor_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#16)
  (0) test_implicit_bit_xor_call

tests -(0)-> <ImplsBitAnd as std::ops::BitAndAssign>::bitand_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#18)
  (0) test_implicit_bit_and_call

tests -(0)-> <ImplsShl as std::ops::ShlAssign>::shl_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#20)
  (0) test_implicit_shl_call

tests -(0)-> <ImplsShr as std::ops::ShrAssign>::shr_assign at tests/ui/call_graph/implicit_bin_op_call.rs:23:9: 23:48 (#22)
  (0) test_implicit_shr_call

targets: 40 total; 40 safe; 0 unsafe (0 tainted)
